lite2/client.go

```diff
@@ -0,0 +1,419 @@
+	}
+}
+
+// SkippingVerification option configures the light client to skip headers as
+// long as {trustLevel} of the old validator set signed the new header. The
+// bisection algorithm from the specification is used for finding the minimal
+// "trust path".
+//
+// trustLevel - fraction of the old validator set (in terms of voting power),
+// which must sign the new header in order for us to trust it. NOTE this only
+// applies to non-adjusted headers. For adjusted headers, sequential
+// verification is used.
+func SkippingVerification(trustLevel cmn.Fraction) Option {
+	if err := ValidateTrustLevel(trustLevel); err != nil {
+		panic(err)
+	}
+	return func(c *Client) {
+		c.verificationMode = skipping
+		c.trustLevel = trustLevel
+	}
+}
+
+// AlternativeSources option can be used to supply alternative providers, which
+// will be used for cross-checking the primary provider.
+func AlternativeSources(providers []provider.Provider) Option {
+	return func(c *Client) {
+		c.alternatives = providers
+	}
+}
+
+// Client represents a light client, connected to a single chain, which gets
+// headers from a primary provider, verifies them either sequentially or by
+// skipping some and stores them in a trusted store (usually, a local FS).
+//
+// Default verification: SkippingVerification(DefaultTrustLevel)
+type Client struct {
+	chainID          string
+	trustingPeriod   time.Duration // see TrustOptions.Period
+	verificationMode mode
+	trustLevel       cmn.Fraction
+
+	// Primary provider of new headers.
+	primary provider.Provider
+
+	// Alternative providers for checking the primary for misbehavior by
+	// comparing data.
+	alternatives []provider.Provider
+
+	// Where trusted headers are stored.
+	trustedStore    store.Store
+	trustedHeader   *types.SignedHeader // H
+	trustedNextVals *types.ValidatorSet // H+1
+
+	logger log.Logger
+}
+
+// NewClient returns a new light client. It returns an error if it fails to
+// obtain the header & vals from the primary or they are invalid (e.g. trust
+// hash does not match with the one from the header).
+//
+// See all Option(s) for the additional configuration.
+func NewClient(
+	chainID string,
+	trustOptions TrustOptions,
+	primary provider.Provider,
+	trustedStore store.Store,
+	options ...Option) (*Client, error) {
+
+	c := &Client{
+		chainID:          chainID,
+		trustingPeriod:   trustOptions.Period,
+		verificationMode: skipping,
+		trustLevel:       DefaultTrustLevel,
+		primary:          primary,
+		trustedStore:     trustedStore,
+		logger:           log.NewNopLogger(),
+	}
+
+	for _, o := range options {
+		o(c)
+	}
+
+	if err := c.initializeWithTrustOptions(trustOptions); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Client) initializeWithTrustOptions(options TrustOptions) error {
+	// 1) Fetch and verify the header.
+	h, err := c.primary.SignedHeader(options.Height)
+	if err != nil {
+		return err
+	}
+
+	// NOTE: Verify func will check if it's expired or not.
```

Comment:
True but maybe Client initialization should fail (? not sure myself)

Reply:
good question